### ISSUE TITLE
Dump unsynced/unsyncable changes to a local file

### ISSUE THREAD

**Issue by amulware** [toggl/mobileapp#7104]
## Overview

In order to make transitioning to the new apps a smooth process for all users, we need to make sure unsynced/unsyncable data gets transitioned correctly.

To do that, we will push that data (and only that data) into a local file, serialised as JSON.

The affected entities are: time entries, projects, clients, tags, since these are the ones the user can create/edit in our apps.

## Format

Everything should be serialised into a single file with the following layout:
```
{ "time_entries" : [], "projects" : [], "clients" : [], "tags" : []}
```
Each of these JSON arrays will contain the unsynced/unsyncable entities of the appropriate nature, exactly the way they would be serialised when sent to our API.

The file is overwritten every time.

## Analytics

I also want an analytics event added to track this, called: `UnsyncedDataDumped` with the following properties:
- `TimeEntries` - number of time entries serialised
- `Projects` - number of time entries serialised
- `Clients` - number of time entries serialised
- `Tags` - number of time entries serialised

## Open questions

A couple of questions remain:

1. How/where to put this file? What is it called?
2. When exactly do we create this file? Every time any of these entities are created/changed? At the beginning or end of sync? This is critical because we need to balance the possible performance impact of the realm queries with making sure the file is always in the correct state.
3. How can we test this?

Please propose answers to these three questions below (discuss if needed) and ask me to sign off on them before starting implementation.

**Comment by juanlaube** [comment]
My thoughts regarding the open questions:

### How/where to put this file? What is it called?
We should use the same filename in both platforms, something like `unsynced-migration-dump.json` is a good name. As for the path, I'd put it directly in the documents folder, in iOS it's the first path in `NSDocumentDirectory` and in Android it's `DIRECTORY_DOCUMENTS`.

### When exactly do we create this file?
I think a new sync step that runs in it's own queue is the best option, this way we can schedule it after pulling, pushing or even cleanup.

### How can we test this?
We could add something that triggers sharing the json file with other apps so we can inspect the contents, maybe triggered by a not-so-obvious gesture like tapping 7 times on the version number or a keyboard shortcut?

**Comment by sofokli1** [comment]
### How/where to put this file? What is it called?

Same as @juanlaube 

### When exactly do we create this file?

Agree with the separate queue, but should happen every time we enter background. I see no point in doing these at any other point since there will always be a chance of syncing the data while the app is still open.

### How can we test this?

If we trigger as I explain above it is easy to test. Otherwise, same way we test loss of workspace access UI. The popup in setting for testing these cases seems like a good place to manually trigger it.

**Comment by pdrj** [comment]
### How/where to put this file? What is it called?

I agree with @juanlaube on the iOS front, since the documents directory is private to the app. On Android however, the proposed place is public and we risk confusion and someone deleting that file. I propose we go with `getFilesDir()` on Android.

### When exactly do we create this file?

I think @sofokli1 has a point about background. I'd like someone more knowledgeable about our sync process to weigh in on this though.

### How can we test this?

We can create an unsyncable TE (extra long description for instance) and then inspect the file's contents. That inspection tool can be accessed via settings, as suggested above.

Should we also validate we're able to restore the TEs and other entities from that file, using the new codebase? I think it is very important to do so, not sure if that should happen as a separate issue though.

**Comment by amulware** [comment]
I agree that we should validate that we have access to the file from the new app, but I'd handle the actual parsing of it only later, once we have the API and syncing implemented, to avoid spending too much time on it now

**Comment by amulware** [comment]
About WHEN: If we want to run it on entering background, we can't have it be in the syncing algorithm, since that might run into background time and get stopped by the OS. I'd say we just trigger this to happen automatically on entering background, running in a separate thread, etc. I don't see right now that anything could go wrong with this (we access realm using proper transactions, as long as we extract all unsynced data in a single transaction as well there shouldn't be any race conditions).

**Comment by sofokli1** [comment]
we should have a sync state (maybe in cleanup) to delete the unsycable file, since after sync it is not needed anymore.

**Comment by daividssilverio** [comment]
### How/where to put this file? What is it called?

I agree with @juanlaube on the iOS front and with @pdrj on the droid front.

### When exactly do we create this file?

Well, since this is going to be part of both new and old apps, a `BackgroundDumpService` sounds like a good idea. We can run in bg at some point in the old app; and run it once in the new apps. Then run the migration to the new databases.

### How can we test this?
We could try to make it unit testable;
Test a bunch of scenarios we can come up with, + all the points above from the others.

**Comment by sofokli1** [comment]
@amulware should development start on this? or do you need more people to reply? (ATM only poor rotation people are looking at this)

**Comment by amulware** [comment]
I'd like this to go ahead, yes. Did we agree on when to trigger this?

**Comment by sofokli1** [comment]
@amulware not 100% agreed. as for iOS:

## How/where to put this file? What is it called?
all agree to call it `unsynced-migration-dump.json` and save it in `NSDocumentDirectory`

## When exactly do we create this file?
- I suggest to create the file when the app goes to background (we can look into background modes for this kind of cases to be triggered when we enter background).
- Delete the file in a new sync cleanup state, since after sync the file is useless.
(there are other some other ideas but this i what I suggest)

## How can we test this?
I see different thoughts on this point and not sure what is the best approach. It is something that can be discussed also after the implementation of the above 2 points.

**Comment by daividssilverio** [comment]
🤔 Let's maybe split this into multiple issues and get this moving?

1. Create the interactor to generate the dump's contents and unit test it.
2. Create the services for each platform, each will call interactor from (1) and save it somewhere. We can decide when to run that service individually for each platform later.

**Comment by amulware** [comment]
> I suggest to create the file when the app goes to background

this makes sense to me, should be the simplest and most reliable.

> Delete the file in a new sync cleanup state, since after sync the file is useless.

Can we ever be sure of this? I rather not mess with this, keep it simpler, and worst case apply a couple of changes twice (unlikely to happen anyways, since the file will probably get overriden with an empty one anyways).

**Comment by arturs105** [comment]
I opened a bunch of smaller issues to split up the work (#7306, #7307, #7308, #7309), so closing this.
